### DISCUSSION TITLE
If a wheel has a .lib.js file we should load it in our loader logic

### DISCUSSION THREAD

**Discussion by szabolcsdombi** [pyodide/pyodide#3945] [Q&A] [answered]
I am interested in getting my ZenGL rendering library working with pyodide out of the box.
I need custom functions provided this way, mostly a one-to-one mapping of the webgl 2 api calls.
I already have such a code available and right now I hack it in with a `pyodide._module.mergeLibSymbols`.

Is the loading of .lib.js files already implemented?

If not, I am interested in doing it, but I might need some guidance.
Or at least I would be willing to test it.

**Comment by ryanking13** [comment]
> Is the loading of .lib.js files already implemented?

I guess you are talking about the follow-up discussed in https://github.com/pyodide/pyodide/discussions/3671#discussioncomment-5364661? No, it is not implemented yet, but I think it is an interesting idea.

> If not, I am interested in doing it, but I might need some guidance.

Yes, it would be really great if you can work on it. Probably the first step we can do is to make an API that loads JavaScript code and merges symbols inside it to Emscripten symbol table.

After that, we can discuss how we should use that API during package loading. Personally, I am interested in using `.pth` file that [the `site` module](https://docs.python.org/3/library/site.html) provides, so that we can stick to the Python standard without making a workaround that works only for Pyodide.

**Reply by szabolcsdombi** [reply]
I was thinking of loading `.lib.js` files automatically. Similar to `.pyi` files.
I think a js function with a specific name should be called from it. Similar to how the Python import machinery works.

```js
// spam.lib.js

function PyInit_spam(pyodide) {
  const wasm = pyodide._module;
  wasm.mergeLibSymbols({
    spam_hello_from_js() {
      console.log('Hello World!');
    },
  });
}
```

```c
// spam.c

extern void spam_hello_from_js();

static PyObject * some_python_accessible_method(PyObject * self, PyObject * args, PyObject * kwargs) {
    spam_hello_from_js();
    Py_RETURN_NONE;
}

extern PyObject * PyInit_spam() {
    return PyModuleDef_Init(&module_def);
}
```

---

PS:

I looked into the `.pth` file, adding anything else but path entries and simple import statements would result in:

```
unsupported 'import' line in ._pth file
```

What you might wanted to refer to was the sitecustomize or the pythonstartup scripts.

**Reply by ryanking13** [reply]
Thanks for the check, then using a `.pth` file doesn't seem to be a good idea.

> I think a js function with a specific name should be called from it.

Yeah, I was thinking of something like `__init__` and `__main__` that would be familiar to Python users. Not a strong opinion, so if you think of a good name, let us know.

**Reply by szabolcsdombi** [reply]
Python supports multiple modules from the same compiled extension by invoking a special method called [`PyInit_<modulename>`](https://docs.python.org/3/extending/extending.html).
For me a `.lib.js` server similar purpose to the `.c` code an extension has.
To extend Python, I write C code and make it available in Python as described in the docs.
To do this on the web I also need `.js` code. (I cannot just use inline `EM_JS()`, see [limitations](https://emscripten.org/docs/compiling/Dynamic-Linking.html#limitations))
I can see the pyodide build size partially depends on similar bindinds, that could be lazy loaded with this method.

I think there is an opportunity for a pure solution to the problem that might be possible to upstream to cpython.
The way I try to use this feature only involves the wasm instance, no proxies are required.
In my example, however, I added the pyodide instance as a parameter and I explain why:

I tried to make a function that has no parameters, and just returns the object to be passed to mergeSymbols:
plenty of problems with this one, the most obvious is the lack of ability to access HEAP.

Then I tried to pass just the wasm instance.
Great, all worked fine until the point I needed to call back into python. (I even tried to use some wasm exported method for the call, but it was a really bad idea)
I actually needed the callbacks for `addEventListener(...)`, `requestAnimationframe(...)` and similar.
It was a clear no-go to assing these from python mostly for boilerplate code and performance purposes.

So my best option now involves an extra script tag for the `.lib.js`. manually calling the `function PyInit_spam(pyodide)` before running the python code. Sometimes I have extra parameters to this method.

I also found a really nice shortcut to this problem.
I serialize the `.lib.js` to my C code before compiling.
after import I can

```py
import js
import pyodide_js

import mymodule  # this module has some imports

init_lib_js_func = js.eval(mymodule.init_lib_js)
init_lib_js_func(pyodide_js._module)

mymodule.do_something_with_the_imported_methods()  # this works
```

**Reply by szabolcsdombi** [reply]
There may be other use cases I do not cover.
Would pure python libraries benefit of these? if they do, whatever returned from the `.lib.js` init script would need to be a proxy. (not true for the C extension case)
Could it be used to just include js libraries like react, vue, ... without returning anything accessible to python?

**Reply by ryanking13** [reply, marked as answer]
> To do this on the web I also need .js code. (I cannot just use inline EM_JS(), see [limitations](https://emscripten.org/docs/compiling/Dynamic-Linking.html#limitations))

I think emscripten now enabled `EM_JS` in side modules (https://github.com/emscripten-core/emscripten/pull/19705).

@sbc100 Could you confirm? I can't find it in the changelog or in the docs.

**Reply by szabolcsdombi** [reply]
I can confirm it is working out of the box with the current emscripten version used by pyodide.

```c++
EM_JS(void, testing_em_js_call_inside_pyodide_extension, (), {
    console.log('this totally worked!!!');
});

static Context * meth_context(PyObject * self, PyObject * args, PyObject * kwargs) {
    testing_em_js_call_inside_pyodide_extension();
    Py_RETURN_NONE;
}
```

![image](https://github.com/pyodide/pyodide/assets/11232402/1b727e3d-1f8d-4de8-92af-f95f1eba3111)

**Reply by szabolcsdombi** [reply]
I picked a very long name to look it up in the imports. I think it is not present there, so it might be emscripten resolves this without adding the function to the import table. I am not 100% sure though.

**Reply by hoodmane** [reply]
A warning though: the way that `EM_JS` functions execute in side modules is different than in the main module, so it is possible to write code that only works in the main module or only works in a side module.

**Comment by sbc100** [comment]
> > To do this on the web I also need .js code. (I cannot just use inline EM_JS(), see [limitations](https://emscripten.org/docs/compiling/Dynamic-Linking.html#limitations))
> 
> I think emscripten now enabled `EM_JS` in side modules ([emscripten-core/emscripten#19705](https://github.com/emscripten-core/emscripten/pull/19705)).
> 
> @sbc100 Could you confirm? I can't find it in the changelog or in the docs.

Yes, EM_JS and EM_ASM now both work in side modules.  Note that in both cases we rely on JS `eval()` here so it perhaps not ideal for all circumstance, but it works.

It looks like you found the specific change and it also looks like I didn't add it to the ChangeLog: https://github.com/emscripten-core/emscripten/pull/19705.

I'm not sure we ever documented the limitation in the first place, but if you see it mentioned anyway that this is *not* possible we should remove that.

**Reply by hoodmane** [reply]
The docs do still say that it's not possible:
> EM_JS functions defined in side modules are not yet supported.

https://emscripten.org/docs/compiling/Dynamic-Linking.html?highlight=em_js#limitations

**Reply by sbc100** [reply]
Oh thanks!  I'll fix that.

**Reply by sbc100** [reply]
https://github.com/emscripten-core/emscripten/pull/20444

**Reply by ryanking13** [reply]
Thanks @sbc100!!

**Comment by szabolcsdombi** [comment]
To load a `.lib.js` alongside the C code is probably a redundant feature as the users can either run js code with pyodide or embed code into C with `EM_JS` and `EM_ASM`.
